### PR TITLE
improve overlapping fields performance

### PR DESCRIPTION
**Pull request by andimarek**
This is a rewrite of the overlapping field can be merged validation, which is quite slow for very large queries.

It is based on the article https://tech.xing.com/graphql-overlapping-fields-can-be-merged-fast-ea6e92e0a01 by @SimonAdameit. Thanks a lot @SimonAdameit for this!

For reference: it is basically the algorithm shown in Listing 10 adapted to the GraphQL Java codebase. I added some basic caching, but it is not equal to Listing 11 in the article.

Performance improvements (on my laptop):

350.048 ± 27.135  ms/op average time vs 35.927 ± 2.913  ms/op

for OverlappingFieldValidationBenchmark.

### CODE REVIEW

File: src/main/java/graphql/validation/rules/OverlappingFieldsCanBeMerged.java [hunk 1]
```diff
@@ -64,187 +68,163 @@ public void leaveSelectionSet(SelectionSet selectionSet) {
...
+        collectFields(fieldMap, fragment.getSelectionSet(), graphQLType, visitedFragmentSpreads);
+    }
 
-        typeA = unwrapAll(typeA);
-        typeB = unwrapAll(typeB);
+    private void collectFieldsForInlineFragment(Map<String, Set<FieldAndType>> fieldMap, Set<String> visitedFragmentSpreads, GraphQLType parentType, InlineFragment inlineFragment) {
+        GraphQLType graphQLType = inlineFragment.getTypeCondition() != null
+                ? TypeFromAST.getTypeFromAST(getValidationContext().getSchema(), inlineFragment.getTypeCondition())
+                : parentType;
+        collectFields(fieldMap, inlineFragment.getSelectionSet(), graphQLType, visitedFragmentSpreads);
```
Comment (bbakerman) [inline src/main/java/graphql/validation/rules/OverlappingFieldsCanBeMerged.java:104]: GraphQLType graphQLType = TypeFromAST.getTypeFromAST(getValidationContext().getSchema(),
                fragment.getTypeCondition());
 
 GraphQLType graphQLType = inlineFragment.getTypeCondition() != null
                ? TypeFromAST.getTypeFromAST(getValidationContext().getSchema(), inlineFragment.getTypeCondition())
                : parentType;
                
                
The above repeated code deserves a method

File: src/main/java/graphql/validation/rules/OverlappingFieldsCanBeMerged.java [hunk 2]
```diff
@@ -270,68 +250,107 @@ private Argument findArgumentByName(String name, List<Argument> arguments) {
...
 
-    private static class FieldPair {
-        final Field field1;
-        final Field field2;
+            FieldAndType that = (FieldAndType) o;
 
-        public FieldPair(Field field1, Field field2) {
-            this.field1 = field1;
-            this.field2 = field2;
+            return field != null ? field.equals(that.field) : that.field == null;
```
Comment (bbakerman) [inline src/main/java/graphql/validation/rules/OverlappingFieldsCanBeMerged.java:348]: This .equals some suspect.

I mean you may only ever compared them into terms of AST Field equality - but the class itself has 3 fields

I just want to make sure this is intended

Comment (andimarek) [reply src/main/java/graphql/validation/rules/OverlappingFieldsCanBeMerged.java:363]: Yes that is intended: the AST field is unique and actually determines the types, meaning we only need to compare the fields.

File: src/main/java/graphql/validation/rules/OverlappingFieldsCanBeMerged.java (hunk 1 above)
Comment (andimarek) [reply src/main/java/graphql/validation/rules/OverlappingFieldsCanBeMerged.java:115]: done ...
just for the record: this is the same code as before: I didn't touch this part for this PR.

On a general note: this is very old code. In general there is a huge overlap in logic between FieldCollector, OverlappingField validation and the normalized field factory.

File: src/main/java/graphql/util/Anonymizer.java [hunk 3]
```diff
@@ -741,17 +741,6 @@ public TraversalControl visitDirective(Directive directive, TraverserContext<Nod
...
-//            public TraversalControl visitStringValue(StringValue node, TraverserContext<Node> context) {
-//                return changeNode(context, node.transform(builder -> builder.value("stringValue" + stringValueCounter.getAndIncrement())));
-//            }
-//
-//            @Override
-//            public TraversalControl visitIntValue(IntValue node, TraverserContext<Node> context) {
-//                return changeNode(context, node.transform(builder -> builder.value(BigInteger.valueOf(intValueCounter.getAndIncrement()))));
-//            }
-//
-
```
Comment (bbakerman) [inline src/main/java/graphql/util/Anonymizer.java:754]: Was this meant?

Comment (andimarek) [reply src/main/java/graphql/util/Anonymizer.java:754]: deleting commented out code .. not directly related to this PR

**Pull request merged**